### PR TITLE
disttask: init capacity and check concurrency using cpu count of managed node

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #49008

Problem Summary:

### What changed and how does it work?
- add cpu_count to mysql.dist_framework_meta(reusing existing ver 180)
- use cpu-count of managed node as capacity when control resource and checking when submitting new task
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
create new and upgrade cluster, make sure new schema works
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/scheduler/nodes.go [hunk 1]
```diff
@@ -110,30 +110,36 @@ func (nm *NodeManager) maintainLiveNodes(ctx context.Context, taskMgr TaskManage
...
-	if newNodes == nil {
-		newNodes = []string{}
+	nodeIDs := make([]string, 0, len(newNodes))
+	var cpuCount int
+	for _, node := range newNodes {
+		nodeIDs = append(nodeIDs, node.ID)
+		if node.CPUCount > 0 {
+			cpuCount = node.CPUCount
+		}
 	}
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/nodes.go:140]: The cpuCount will be set equal to the last node, is it expected?

File: pkg/executor/set.go [hunk 2]
```diff
@@ -172,8 +172,11 @@ func (e *SetExecutor) setSysVariable(ctx context.Context, name string, v *expres
 			dom := domain.GetDomain(e.Ctx())
 			config.GetGlobalConfig().Instance.TiDBServiceScope = valStr
 			serverID := disttaskutil.GenerateSubtaskExecID(ctx, dom.DDL().GetID())
-			_, err = e.Ctx().(sqlexec.SQLExecutor).ExecuteInternal(ctx,
-				`replace into mysql.dist_framework_meta values(%?, %?, DEFAULT)`, serverID, valStr)
+			taskMgr, err := storage.GetTaskManager()
+			if err != nil {
+				return err
+			}
+			return taskMgr.StartManagerSession(ctx, e.Ctx(), serverID, valStr)
```
Comment (ywqzzy) [inline pkg/executor/set.go:179]: If the sysvar is not set, the node won't be used for disttask.
Is it exepected?

File: pkg/disttask/framework/storage/task_table_test.go [hunk 3]
```diff
@@ -37,6 +39,16 @@ func TestMain(m *testing.M) {
 	goleak.VerifyTestMain(m, opts...)
 }
 
+func GetCPUCountOfManagedNodes(ctx context.Context, taskMgr *TaskManager) (int, error) {
```
Comment (ywqzzy) [inline pkg/disttask/framework/storage/task_table_test.go:42]: move to util?

File: pkg/disttask/framework/scheduler/nodes.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/nodes.go:140]: we assume them isomorphic, first/last makes no difference

File: pkg/executor/set.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/executor/set.go:179]: why? i just replace it from explicit sql to task_table api

Comment (ywqzzy) [reply pkg/executor/set.go:179]: Okay, I missunderstood the code

File: pkg/disttask/framework/storage/task_table_test.go (hunk 3 above)
Comment (D3Hunter) [reply pkg/disttask/framework/storage/task_table_test.go:42]: let's do it when it's needed

File: pkg/disttask/framework/scheduler/slots.go [hunk 4]
```diff
@@ -47,10 +50,7 @@ type taskStripes struct {
 // quota to subtask, but subtask can determine what to conform.
 type slotManager struct {
 	// Capacity is the total number of slots and stripes.
-	// TODO: we assume that all nodes managed by dist framework are isomorphic,
-	// but dist owner might run on normal node where the capacity might not be
-	// able to run any task.
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/slots.go:52]: Add some comments of isomorphic

File: pkg/disttask/framework/storage/task_table.go [hunk 5]
```diff
@@ -1249,36 +1259,78 @@ func (stm *TaskManager) TransferTasks2History(ctx context.Context, tasks []*prot
...
-	nodes := make([]string, 0, len(rs))
+	nodes := make([]proto.ManagedNode, 0, len(rs))
 	for _, r := range rs {
-		nodes = append(nodes, r.GetString(0))
+		role := r.GetString(1)
+		nodes = append(nodes, proto.ManagedNode{
+			ID:       r.GetString(0),
+			Role:     role,
+			CPUCount: int(r.GetInt64(2)),
+		})
```
Comment (ywqzzy) [inline pkg/disttask/framework/storage/task_table.go:1314]:
```suggestion
		nodes = append(nodes, proto.ManagedNode{
			ID:       r.GetString(0),
			Role:     r.GetString(1),
			CPUCount: int(r.GetInt64(2)),
		})
```

**Review by ywqzzy — APPROVED** [review]
Rest LGTM

File: pkg/disttask/framework/scheduler/slots.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/slots.go:52]: see comments in slotManager

File: pkg/disttask/framework/scheduler/nodes.go (hunk 1 above)
Comment (okJiang) [reply pkg/disttask/framework/scheduler/nodes.go:140]: can add a comment~

File: pkg/disttask/framework/storage/task_table.go [hunk 6]
```diff
@@ -1249,36 +1259,77 @@ func (stm *TaskManager) TransferTasks2History(ctx context.Context, tasks []*prot
...
+		return err2
+	})
+	return nodes, err
+}
+
+func (*TaskManager) getAllNodesWithSession(ctx context.Context, se sessionctx.Context) ([]proto.ManagedNode, error) {
+	rs, err := ExecSQL(ctx, se, `
+		select host, role, cpu_count
+		from mysql.dist_framework_meta
+		where role = 'background' or role = ''
```
Comment (okJiang) [inline pkg/disttask/framework/storage/task_table.go:1302]: how about removing this?

**Review by okJiang — APPROVED** [review]

**Comment by D3Hunter** [comment]
/hold 

ci fail

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by D3Hunter** [comment]
ci fail issue, will fix it later, https://github.com/pingcap/tidb/issues/49950

**Review by yudongusa — APPROVED** [review]
LGTM for now, but @Benjamin2037 please consider the version rollback design.

**Pull request merged**